File: src/main/scala/ieee80211/IEEE80211.scala
package modem

/**
 * Sourced from https://github.com/grebe/ofdm/src/main/scala/ieee80211/IEEE80211.scala
 */

import breeze.linalg.DenseVector
import breeze.math.Complex
import breeze.numerics.{cos, sin, sqrt}
import breeze.signal.{iFourierShift, iFourierTr}

object IEEE80211 {
  val zero = Complex(0, 0)
  val one  = Complex(1, 0)
  val pp   = Complex(1, 1)
  val pm   = Complex(1, -1)
  val mp   = Complex(-1, 1)
  val mm   = Complex(-1, -1)

  // indexed from bin -26 -> bin 26
  val stfFreq = DenseVector(
    zero, zero, zero, zero, mm,   zero, zero, zero,
    mm,   zero, zero, zero, pp,   zero, zero, zero,
    pp,   zero, zero, zero, pp,   zero, zero, zero,
    pp,   zero, zero, zero, zero, zero, zero, zero,
    zero, zero, zero, zero, zero, zero, zero, zero,
    pp,   zero, zero, zero, mm,   zero, zero, zero,
    pp,   zero, zero, zero, mm,   zero, zero, zero,
    mm,   zero, zero, zero, pp,   zero, zero, zero
  ) map { x => sqrt(13.0 / 6.0) * x }
  val stf64 = iFourierTr(stfFreq)

  val stf = DenseVector.vertcat(stf64, stf64, stf64).toArray.slice(0, 160)

  val ltfFreq = DenseVector(
    zero,
    one, -one, -one,  one,  one, -one,  one, -one,
    one, -one, -one, -one, -one, -one, one,  one,
    -one, -one,  one, -one,  one, -one,  one,  one,
    one,  one,

    zero, zero, zero, zero, zero,
    zero, zero, zero, zero, zero, zero,

    one, one,
    -one, -one, one, one, -one, one, -one, one,
    one, one, one, one, one, -one, -one, one,
    one, -one, one, -one, one, one, one, one

  )

  val ltf64 = iFourierTr(ltfFreq)

  val ltf = DenseVector.vertcat(ltf64, ltf64, ltf64).toArray.slice(32,160+32)

  def addCFO(in: Seq[Complex], cfo: Double = 0.0, sampleRate: Double = 20.0e6): Seq[Complex] = {
    val coeff = 2 * math.Pi * cfo / sampleRate
    in.zipWithIndex.map { case (samp, idx) =>
      val rotation = Complex(cos(coeff * idx), sin(coeff * idx))
      rotation * samp
    }
  }

  def main(args: Array[String]): Unit = {
    println(stf64)
    println(s"Length is ${stf64.size}")

    println(s"LTF is $ltf64")
  }
}

File: src/model/IEEE80211.py
##
# Sourced from https://github.com/grebe/ofdm/src/main/scala/ieee80211/IEEE80211.scala
#

import numpy as np


zero = 0.+0.j
one  = 1.+0.j
pp   = 1.+1.j
pm   = 1.-1.j
mp   = -1.+1.j
mm   = -1.-1.j


class IEEE80211:
    # indexed from bin -26 -> bin 26
    stfFreq = np.array([
            zero, zero, zero, zero, mm,   zero, zero, zero,
            mm,   zero, zero, zero, pp,   zero, zero, zero,
            pp,   zero, zero, zero, pp,   zero, zero, zero,
            pp,   zero, zero, zero, zero, zero, zero, zero,
            zero, zero, zero, zero, zero, zero, zero, zero,
            pp,   zero, zero, zero, mm,   zero, zero, zero,
            pp,   zero, zero, zero, mm,   zero, zero, zero,
            mm,   zero, zero, zero, pp,   zero, zero, zero
    ]) * np.sqrt(13.0 / 6.0)
    stf64 = np.fft.ifft(stfFreq)

    stf = np.concatenate([stf64, stf64, stf64])[0:160]

    ltfFreq = np.array([
            zero,
            one, -one, -one,  one,  one, -one,  one, -one,
            one, -one, -one, -one, -one, -one, one,  one,
            -one, -one,  one, -one,  one, -one,  one,  one,
            one,  one,

            zero, zero, zero, zero, zero,
            zero, zero, zero, zero, zero, zero,

            one, one,
            -one, -one, one, one, -one, one, -one, one,
            one, one, one, one, one, -one, -one, one,
            one, -one, one, -one, one, one, one, one
    ])

    ltf64 = np.fft.ifft(ltfFreq)

    ltf = np.concatenate([ltf64, ltf64, ltf64])[32:160+32]

    def __init__(self):
        pass

    def addCFO(self, iq, cfo=0.0, sampleRate=20.0e6):
        coeff = 2 * np.pi * cfo / sampleRate
        return map(lambda idx, samp: (np.cos(coeff * idx) + np.sin(coeff * idx)*1j) * samp, enumerate(iq))

    def main(self, args):
        print(self.stf64)
        print("Length is {}".format(self.stf64.size))

        print("LTF is {}".format(self.ltf64))

if __name__ == "__main__":
    IEEE80211().main([])
